### PR TITLE
Add one Jaeger span per relay parent

### PR DESCRIPTION
**Pull request by bkchr**
This adds one Jaeger span per relay parent, instead of always creating
new spans per relay parent. This should improve the UI view, because
subsystems are now grouped below one common span.

### CODE REVIEW

File: node/subsystem/src/lib.rs [hunk 1]
```diff
@@ -48,18 +49,21 @@ const ACTIVE_LEAVES_SMALLVEC_CAPACITY: usize = 8;
...
 ///
 /// Note that the activated and deactivated fields indicate deltas, not complete sets.
-#[derive(Clone, Debug, Default, Eq)]
+#[derive(Clone, Default)]
 pub struct ActiveLeavesUpdate {
-	/// New relay chain block hashes of interest.
-	pub activated: SmallVec<[Hash; ACTIVE_LEAVES_SMALLVEC_CAPACITY]>,
+	/// New relay chain block hashes of interest and their associated [`JaegerSpan`].
+	///
+	/// NOTE: Each span should only kept active as long as the leave is considered active and should be dropped
```
Comment (rphmeier) [inline node/subsystem/src/lib.rs:56]: typo: leaf

File: node/subsystem/src/lib.rs [hunk 2]
```diff
@@ -48,18 +49,21 @@ const ACTIVE_LEAVES_SMALLVEC_CAPACITY: usize = 8;
...
+	/// when the leave is deactivated.
+	pub activated: SmallVec<[(Hash, Arc<JaegerSpan>); ACTIVE_LEAVES_SMALLVEC_CAPACITY]>,
```
Comment (rphmeier) [inline node/subsystem/src/lib.rs:58]: Yeah, adding the span to the leaf is a good idea

File: node/jaeger/src/lib.rs [hunk 3]
```diff
@@ -102,6 +113,50 @@ impl JaegerConfigBuilder {
...
+/// A special "per leave span".
+///
+/// Essentially this span wraps two spans:
+///
+/// 1. The span that is created per leave in the overseer.
+/// 2. Some child span of the per-leave span.
+///
+/// This just works as auxiliary structure to easily store both.
+#[derive(Debug)]
+pub struct PerLeaveSpan {
```
Comment (rphmeier) [inline node/jaeger/src/lib.rs:125]: should be PerLeafSpan

File: node/jaeger/src/lib.rs [hunk 4]
```diff
@@ -102,6 +113,50 @@ impl JaegerConfigBuilder {
...
+	span: JaegerSpan,
+}
+
+impl PerLeaveSpan {
+	/// Creates a new instance.
+	///
+	/// Takes the `leave_span` that is created by the overseer per leave and a name for a child span.
+	/// Both will be stored in this object, while the child span is implicitly accessible by using the
+	/// [`Deref`](std::ops::Deref) implementation.
+	pub fn new(leave_span: Arc<JaegerSpan>, name: impl Into<String>) -> Self {
```
Comment (rphmeier) [inline node/jaeger/src/lib.rs:136]: leave -> leaf

File: node/jaeger/src/lib.rs [hunk 5]
```diff
@@ -102,6 +113,50 @@ impl JaegerConfigBuilder {
 	}
 }
 
+/// A special "per leave span".
```
Comment (rphmeier) [inline node/jaeger/src/lib.rs:116]: same

**Comment by rphmeier** [comment]
Seems fine to me except for naming

File: node/core/backing/src/lib.rs [hunk 6]
```diff
@@ -926,6 +926,7 @@ impl util::JobTrait for CandidateBackingJob {
 	#[tracing::instrument(skip(keystore, metrics, rx_to, tx_from), fields(subsystem = LOG_TARGET))]
 	fn run(
 		parent: Hash,
+		span: Arc<JaegerSpan>,
```
Comment (coriolinus) [inline node/core/backing/src/lib.rs:929]: Could add this to the `skip` section of the tracing line.

File: node/core/bitfield-signing/src/lib.rs [hunk 7]
```diff
@@ -218,14 +218,15 @@ impl JobTrait for BitfieldSigningJob {
 	#[tracing::instrument(skip(keystore, metrics, _receiver, sender), fields(subsystem = LOG_TARGET))]
 	fn run(
 		relay_parent: Hash,
+		span: Arc<JaegerSpan>,
```
Comment (coriolinus) [inline node/core/bitfield-signing/src/lib.rs:221]: Likewise, I think it likely that the span can be skipped from the tracing here without sacrificing useful debug information.

File: node/core/provisioner/src/lib.rs [hunk 8]
```diff
@@ -143,6 +142,7 @@ impl JobTrait for ProvisioningJob {
 	#[tracing::instrument(skip(_run_args, metrics, receiver, sender), fields(subsystem = LOG_TARGET))]
 	fn run(
 		relay_parent: Hash,
+		span: Arc<JaegerSpan>,
```
Comment (coriolinus) [inline node/core/provisioner/src/lib.rs:145]: skip span please

File: node/subsystem/src/lib.rs [hunk 9]
```diff
@@ -48,18 +49,21 @@ const ACTIVE_LEAVES_SMALLVEC_CAPACITY: usize = 8;
...
 ///
 /// Note that the activated and deactivated fields indicate deltas, not complete sets.
-#[derive(Clone, Debug, Default, Eq)]
+#[derive(Clone, Default)]
 pub struct ActiveLeavesUpdate {
-	/// New relay chain block hashes of interest.
-	pub activated: SmallVec<[Hash; ACTIVE_LEAVES_SMALLVEC_CAPACITY]>,
+	/// New relay chain block hashes of interest and their associated [`JaegerSpan`].
+	///
+	/// NOTE: Each span should only kept active as long as the leaf is considered active and should be dropped
```
Comment (ordian) [inline node/subsystem/src/lib.rs:56]:
```suggestion
	/// NOTE: Each span should only be kept active as long as the leaf is considered active and should be dropped
```

**Review by ordian — APPROVED** [review]
Looks good modulo unresolved comments.

**Review by coriolinus — APPROVED** [review]

**Review by drahnr — APPROVED** [review]

**Pull request merged**